### PR TITLE
Themes: Add verticals description to SSR'd HTML

### PR DESCRIPTION
**Pull request by ockham**
This PR:
* Enables the data layer middleware on the server (yes, works fine :slightly_smiling_face: )
* Updates the theme filters schema to reflect the API
* Add a `requestThemeFilters` action and a number of related selectors (plus unit tests)...
* ... which are used by the `themes` section's controller to fetch themes before server-side rendering the showcase
* Changes the `ThemeShowcase` component to use the right data for the `description` `meta` tag.

To test:
* Restart Calypso: `DEBUG="calypso:themes" make run`
* While logged out, verify in Page Source that there's a specific `description` `meta` in the following cases (you need to land on these pages directly):
  * Vertical:  http://calypso.localhost:3000/themes/photography
  * Single Filter: http://calypso.localhost:3000/themes/filter/minimal
  * Tier: http://calypso.localhost:3000/themes/premium
  * Generic: http://calypso.localhost:3000/themes/
* Note that not all verticals and filters have a description; in those cases, we fall back to the next lower level (i.e. no verticals -> filter; neither verticals nor filters -> tier).
* Multiple filters are equivalent to no filters at all (i.e. the next lower level is used).
* Verify that after the first hit, theme filters are cached (watch out for `found theme filters in cache` in your node console).

---

(Original description:)

Background (from Slack):

> @dmsnell I have a conceptual question about the data/HTTP middleware. As you know, the `themeFilters` reducer is implemented to use that. For server-side rendering, we require data fetched by the corresponding middleware. With other network fetches implemented as `requestSomething` thunks, we usually chain `.then( next )` in the corresponding `page.js` controller in order to hold execution until we've received the required data. Q is, how would I go about achieving the same effect with the data/HTTP mw, where all consumer-facing actions are synchronous?

FYI @dmsnell @blowery @gwwar 

Edit: props @blowery for the idea how to have the router's middleware wait for theme filters to be fetched.

https://github.com/Automattic/wp-calypso/pull/10988 might be a reasonable future step to simplify data caching.

### CODE REVIEW

**Comment by ockham** [comment]
FYI @seear @folletto @budzanowski

**Comment by dmsnell** [comment]
(from Slack)

@ockham how do we know the last promise has returned?

I assume there are a bag of N promises fetching data? I'm guessing that right now we still make some kind of decision that says, "all data requests which will occur have started" but I don't know how we track those. how do we know that the last promise has completed? or can we pause the render of a component from inside that component until it gets the resolved promise?

as @blowery mentioned it’s an unaddressed question though I think we could have at least a reasonable solution with the data layer without much augmentation. this depends on how we currently do it and what our time budget is

suppose we need to make twenty network requests before we can render the page. when running server-side we add a new piece in the HTTP middleware which simply tracks outgoing requests and incoming responses.

if we don’t know when the requests will stop we could do something like

```if no more requests come within 5ms consider the request phase done
count the number of outbound requests
set a timeout at 50ms
once the last request returns or the timeout satisfies
fire a new Redux action INITIAL_DATA_FETCH_COMPLETE
```

now we simply subscribe to the store to watch that action fly by and when it does we know we can rerender the last time and ship

**Comment by ockham** [comment]
> we know we can rerender the last time and ship

On the server, we only get one shot to render, so we need to have our data ready before rendering for the _first_ (and only) time.

Ideally, the interface to this would allow us to specify our data needs (at component level), and only render once they're available...

**Comment by dmsnell** [comment]
> On the server, we only get one shot to render, so we need to have our data ready before rendering for the first (and only) time.
> 
> Ideally, the interface to this would allow us to specify our data needs (at component level), and only render once they're available...

how do we currently _**a**_ know that we need a specific piece of data and _**b**_ block render until we have it?

again, if we know how to make those data requests (we could even be naïve and record a session to grab the request actions) then we could push those through the data layer and wait for the last one to come back or timeout then render with the new tree. seems in some ways easier than I thought if we know how to figure out what data we need.

**Comment by ockham** [comment]
> how do we currently _**a**_ know that we need a specific piece of data and _**b**_ block render until we have it?

With thunk based network fetches, the answer to both is that we [chain `.then( next )`](https://github.com/Automattic/wp-calypso/blob/54a5e07b725aeb31685952f56731c20abe6f6130/client/my-sites/themes/controller.jsx#L133-L143) to the corresponding `dipatch` in the data fetching router middleware.

**Comment by folletto** [comment]
Tested in Chrome / CURL:

### /themes/photography

```<title>WordPress Themes — WordPress.com</title>```

```<meta name="description" property="og:description" content="A WordPress photography theme is great place to showcase your photography work on the web. Our photography themes include elegant image-portfolios features and ways to &lt;a href=&quot;http://en.blog.wordpress.com/2011/11/08/new-photo-carousel/&quot;&gt;browse galleries&lt;/a&gt; that are both immersive and gorgeous.">```

### /themes/filter/minimal

```<title>WordPress Themes — WordPress.com</title>```

```<meta name="description" property="og:description" content="Whether you're minimalist at heart, like keeping things clean, or just want to focus — you'll find a robust collection of minimal WordPress themes here. Achieve blogging zen and cut straight to the good stuff by choosing one of these gorgeously simple WordPress themes.">```

### /themes/premium

```<title>Premium WordPress Themes — WordPress.com</title>```

```<meta name="description" property="og:description" content="Discover Premium WordPress Themes on the WordPress.com Theme Showcase.">```


### /themes

```<title>WordPress Themes — WordPress.com</title>```

```<meta name="description" property="og:description" content="Beautiful, responsive, free and premium WordPress themes for your photography site, portfolio, magazine, business website, or blog.">```

---

_I know this PR isn't about titles, but I reported these just to keep them visible._

1. I'm dubious about URLs in there, but we can edit them out as this PR makes them driven from the DB.
2. They seem working well

The only ? I have is this:

> Note that not all verticals and filters have a description; in those cases, we fall back to the next lower level (i.e. no verticals -> filter; neither verticals nor filters -> tier).

Can we instead use mad libs there instead of fallbacks if there are no descriptions coming from the DB?

**Comment by seear** [comment]
Instead of these new selectors, should we maybe be adding functions to the [existing library](https://github.com/Automattic/wp-calypso/blob/master/client/my-sites/themes/theme-filters.js), so we don't have related functions spread out?

File: client/state/selectors/find-theme-filter-term.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+import { getThemeFilters, getThemeFilterTerm } from './';
+
+/**
+ * Returns a theme filter term object that corresponds to a given filter term slug
+ *
+ * @param  {Object}  state  Global state tree
+ * @param  {String}  search The term to search for
+ * @return {Object}         A filter term object
+ */
+export default function findThemeFilterTerm( state, search ) {
```
Comment (seear) [inline client/state/selectors/find-theme-filter-term.js]: In the fillter lib we build a [table of terms to tax](https://github.com/Automattic/wp-calypso/blob/bbc3832a80980a0ac37379a0bb73d66acbf3806e/client/my-sites/themes/theme-filters.js#L194) (would be more data now we've added to the endpoint), that would make this a simple lookup in the table, I think?

**Comment by seear** [comment]
Let's not let any decisions about selectors vs lib delay this PR, though, because we can always tidy up when we come to use the API for the filtering in the search box.

**Comment by ockham** [comment]
As discussed in Slack, merging this so we get the ball rolling. Let's iterate on the mad-libs, and the library approach/data representation.

**Pull request merged**